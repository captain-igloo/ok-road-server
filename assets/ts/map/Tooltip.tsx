import { LeafletMouseEvent } from 'leaflet';
import * as React from 'react';
import { useMap, useMapEvents } from 'react-leaflet';

import { useAppDispatch } from '../store';
import HtmlLayer from './HtmlLayer';
import { setTooltip, Tooltip } from './slice';

export default function (props: { tooltip: Tooltip }) {
    const { tooltip } = props;
    const map = useMap();
    const [htmlLayer, setHtmlLayer] = React.useState<HtmlLayer>();

    const dispatch = useAppDispatch();

    const mouseHandler = React.useMemo(() => (e: LeafletMouseEvent) => {
        dispatch(setTooltip({
            position: {
                lat: e.latlng.lat,
                lng: e.latlng.lng,
            },
        }));
    }, []);

    useMapEvents({
        click: mouseHandler,
        mousemove: mouseHandler,
    });

    React.useEffect(() => {
        const newHtmlLayer = new HtmlLayer({ className: 'map-tooltip' });
        newHtmlLayer.addTo(map);
        setHtmlLayer(newHtmlLayer);
        return () => {
            newHtmlLayer.remove();
        };
    }, []);

    React.useEffect(() => {
        if (htmlLayer) {
            htmlLayer.setPosition(tooltip.position);
            htmlLayer.setHtml(tooltip.text);
        }
    }, [htmlLayer, tooltip]);

    return null;
}
