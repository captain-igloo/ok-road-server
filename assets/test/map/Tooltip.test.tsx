import '../setup';

import {
    describe,
    expect,
    jest,
    test,
} from '@jest/globals';
import { render } from '@testing-library/react';
import * as React from 'react';
import { useMap } from 'react-leaflet';

import Tooltip from '../../ts/map/Tooltip';

describe('Tooltip component', () => {
    test('Tooltip should render properly', () => {
        const addLayer = jest.fn();
        (useMap as any).mockImplementation(() => ({
            addLayer,
        }));
        const { container } = render(
            <Tooltip tooltip={{} as any}/>
        );
        expect(container).toMatchSnapshot();
    });
});
