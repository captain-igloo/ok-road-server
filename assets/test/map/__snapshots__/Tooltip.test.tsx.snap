// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip component Tooltip should render properly 1`] = `<div />`;
